### PR TITLE
Implement validity checks

### LINKED ISSUE #301: Missing check for object correctness invariants
Rust requires that all values have to be valid when assigned to or read from a place, passed to a function/primitive operation or returned from a function/primitive operation.  It is UB to produce an invalid bit-pattern of a given type.  We do not currently have any checks for this in Kani.

## Likelihood:

If code contains this bug, Kani will not detect it. We do not have any data as to how often this occurs in practice.

## Mitigation: 

Warn users that this class of UB is outside the current scope of Kani.

## Path to soundness:

Add a check for this.

### PR DESCRIPTION
**Pull request by celinval**
This is still incomplete, but hopefully it can be merged as an unstable feature. I'll publish an RFC shortly.

This instruments the function body with assertion checks to see if users are generating invalid values. This covers:
  - Union access
  - Raw pointer dereference
  - Transmute value
  - Field assignment of struct with invalid values
  - Aggregate assignment

Things not covered today should trigger ICE or a delayed verification failure due to unsupported feature.

## Design

This change has two main design changes which are inside the new `kani_compiler::kani_middle::transform` module:
  1- Instance body should now be retrieved from the `BodyTransformation` structure. This structure will run transformation passes on instance bodies (i.e.: monomorphic instances) and cache the result.
  2- Create a new transformation pass that instruments the body of a function for every potential invalid value generation.
  3- Create a body builder which contains all elements of a function body and mutable functions to modify them accordingly.

## Call-outs
  1. I am planning to add expected tests in a follow up PR.
  2. This PR is currently blocked since it requires toolchain nightly-2024-03-15.
  3. One potential improvement for later is to replace most of the build check logic by an internal Kani library function that handle the comparison.

Related to #2998
Fixes #301 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/kani/ValidValues/write_invalid.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+//! Check that Kani can identify UB after writing an invalid value.
+//! Writing invalid bytes is not UB as long as the incorrect value is not read.
+//! However, we over-approximate for sake of simplicity and performance.
+
+use std::num::NonZeroU8;
+
+#[kani::proof]
+#[kani::should_panic]
+pub fn write_invalid_bytes_no_ub() {
```
Comment (zhassan-aws) [inline tests/kani/ValidValues/write_invalid.rs:13]: This test seems to be subsumed by the next one.

**Review by zhassan-aws — COMMENTED** [review]
The first round of comments. I haven't finished reviewing the entire PR yet.

File: tests/kani/ValidValues/custom_niche.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        Rating { stars }
+    }
+}
+
+#[kani::proof]
+#[kani::should_panic]
+pub fn check_new_with_ub() {
+    assert_eq!(Rating::new(10), None);
+}
+#[test]
```
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/ValidValues/custom_niche.rs:42]: No. Thanks

File: tests/kani/ValidValues/custom_niche.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    ];
+    kani::concrete_playback_run(concrete_vals, check_unchecked_new_ub);
+}
+
+#[kani::proof]
+#[kani::should_panic]
+pub fn check_new_with_ub_limits() {
+    let stars = kani::any_where(|s: &u8| *s == 0 || *s > 5);
+    let _ = Rating::new(stars);
+    unreachable!("Call to new should always fail");
```
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/ValidValues/custom_niche.rs:68]: I will remove that. I was planning to make this an expected test, but I'm waiting for some feedback on the RFC

**Review by zhassan-aws — COMMENTED** [review]
Some nits

File: kani-compiler/src/args.rs [hunk 4]
```diff
@@ -71,4 +71,18 @@ pub struct Arguments {
     #[clap(long)]
     /// A legacy flag that is now ignored.
     goto_c: bool,
+    /// Enable specific checks.
+    #[clap(long, conflicts_with = "check_all")]
+    pub ub_check: Vec<ExtraChecks>,
+    /// Enable all extra checks.
+    #[clap(long)]
+    pub check_all: bool,
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/args.rs:79]: Actually, I think I'll just remove this for now. I'll leave this for after we decide on #3092

File: kani_metadata/src/unstable.rs [hunk 5]
```diff
@@ -84,6 +84,8 @@ pub enum UnstableFeature {
     FunctionContracts,
     /// Memory predicate APIs.
     MemPredicates,
+    /// Automatically check for that all values checks.
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani_metadata/src/unstable.rs:87]: it's incomplete... let me fix that

File: tests/kani/ValidValues/write_invalid.rs (hunk 1 above)
Comment (celinval) [reply tests/kani/ValidValues/write_invalid.rs:13]: I actually would prefer keeping it. In an ideal world, this test should not panic. This test shows that the UB detection for this case today is an over-approximation. I could turn it into a fixme test, but it's not really a bug at this point, it is a decision made by design.

File: kani-compiler/src/kani_middle/transform/mod.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    pub fn dummy() -> Self {
+        BodyTransformation { opt_passes: vec![], inst_passes: vec![], cache: Default::default() }
+    }
+
+    pub fn body(&mut self, tcx: TyCtxt, instance: Instance) -> Option<Body> {
+        match self.cache.get(&instance) {
+            Some(TransformationResult::Modified(body)) => Some(body.clone()),
+            Some(TransformationResult::NotModified) => instance.body(),
+            None => {
+                let mut body = instance.body()?;
```
Comment (zhassan-aws) [inline kani-compiler/src/kani_middle/transform/mod.rs:69]: Is there a way this can return `None`?

**Review by zhassan-aws — APPROVED** [review]
Awesome work, @celinval!

File: kani-compiler/src/kani_middle/transform/mod.rs (hunk 6 above)
Comment (celinval) [reply kani-compiler/src/kani_middle/transform/mod.rs:69]: There are cases where instance.body() does return `None`, but I believe all callers to this function unwrap the result. So I could remove the `Option`.

Comment (zhassan-aws) [reply kani-compiler/src/kani_middle/transform/mod.rs:69]: That would make sense.

**Pull request merged**